public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryEditorSelector.tsx

```diff
@@ -49,6 +66,18 @@ export const PromQueryEditorSelector = React.memo<PromQueryEditorProps>((props)
 
   return (
     <>
+      <ConfirmModal
+        isOpen={parseModalOpen}
+        title="Query parsing"
+        body="There were errors while trying to parse the query. Continuing to visual builder may loose some parts of the query."
+        confirmText="Continue"
+        onConfirm={() => {
+          setVisualQuery(pendingChange!.visualQuery!);
```

Comment:
I've had various devs ask about the `!` operator. So this as much asking about what the standard really is. If `setVisualQuery` must have a non-null parameter, should be shy away from this operator and be very explicit?

Reply:
yeah I don't like using the `!` either tbh but this is such a case when I am on the fence. The problem is `pendingChange` can be `undefined` and if it is defined then it is a `promQuery` which can have `visualQuery` undefined. But this modal is shown only in the case when the pending change is set and has a visual query.

On the other hand there can be some programmatic error that will break this relation (because it is not check in compile time) and so maybe having a check here like:
```
if (pendingChange?.visualQuery) {
  //...setState
}
```
would make sense. But the caveat here is that if I do that and the relation at some point breaks it will just silently do nothing which IMHO is even worse.

Reply:
Ideally you would represent these state changes directly in types somehow so the open/closed modal state is tied to the pendingChange state so they cannot be in a state where modal is open and pending state is undefined but here I am not sure how to do that without still having to check for undefined inside the onConfirm (because the execution of that handler is just at different time than rendering so there can be state change in between and typescript will still think something could have changed the pendingChange to `undefined` and it would be technically correct).